src/vsphere_cpi/lib/cloud/vsphere/attach_tag_to_vm.rb

```diff
@@ -0,0 +1,134 @@
+require 'cloud/vsphere/logger'
+require 'vsphere-automation-cis'
+require 'vsphere-automation-vcenter'
+
+module  VSphereCloud
+  module TaggingTag
+    class AttachTagToVm
+      include Logger
+
+      def initialize(api_client)
+        @api_client = api_client
+      end
+
+      private def tag_assoc_info
+        @tag_assoc_info ||= VSphereAutomation::CIS::CisTaggingTagAssociationAttach.new
+      end
+
+      private def multi_tag_assoc_info
+        @multi_tag_assoc_info ||= VSphereAutomation::CIS::CisTaggingTagAssociationAttachMultipleTagsToObject.new
+      end
+
+      private def tag_assoc_api
+        @tag_assoc_api ||= VSphereAutomation::CIS::TaggingTagAssociationApi.new(@api_client)
+      end
+
+      private def tagging_tag_api_instance
+        @tagging_tag_api_instance ||= VSphereAutomation::CIS::TaggingTagApi.new(@api_client)
+      end
+
+      private def tagging_category_api_instance
+        @tagging_category_api_instance ||= VSphereAutomation::CIS::TaggingCategoryApi.new(@api_client)
+      end
+
+      def verify_category_id( config_category, category_ids_on_host )
+        category_ids_on_host.each do |category_id|
```

Comment:
Instead of passing this in the function, this can be retrieved here directly. What benefit do you get in making this a param in callee function?

Reply:
I made it this way to keep all warning inside of function attach_tags. If this is empty array, a warning is raised in  attach_tags. If we do not need warnings, I will change.